js/data/array_utils.js

```diff
@@ -78,21 +79,36 @@ function setDataByKeyMapValue(array, key, data) {
     }
 }
 
-function applyBatch(keyInfo, array, batchData, groupCount, useInsertIndex) {
+function applyBatch(keyInfo, array, batchData, groupCount, useInsertIndex, immutable) {
+    const isDataImmutable = isDefined(immutable) && immutable === true;
+    const resultItems = isDataImmutable ? [...array] : array;
+
     batchData.forEach(item => {
-        const items = item.type === 'insert' ? array : getItems(keyInfo, array, item.key, groupCount);
+        const items = item.type === 'insert' ? resultItems : getItems(keyInfo, resultItems, item.key, groupCount);
 
-        generateDataByKeyMap(keyInfo, items);
+        !isDefined(immutable) && generateDataByKeyMap(keyInfo, items);
 
         switch(item.type) {
-            case 'update': update(keyInfo, items, item.key, item.data, true); break;
+            case 'update': update(keyInfo, items, item.key, item.data, true, isDataImmutable); break;
             case 'insert': insert(keyInfo, items, item.data, useInsertIndex && isDefined(item.index) ? item.index : -1, true); break;
             case 'remove': remove(keyInfo, items, item.key, true); break;
         }
     });
+    return resultItems;
+}
+
+function applyChanges(data, changes, options) {
+    const { keyExpr = 'id', immutable = true } = isDefined(options) && isPlainObject(options) ? options : {};
```

Comment:
You can define default value for the options and remove all these checks. 
```
function applyChanges(data, changes, options = {}) {
    const { keyExpr = 'id', immutable = true } = options;
```

Reply:
makes sense:)